db/database.go

```diff
@@ -2331,6 +2336,22 @@ func (dbc *DatabaseContext) GetSingleDatabaseCollection() *DatabaseCollection {
 	return dbc.singleCollection
 }
 
+func validateMetadataStore(ctx context.Context, metadataStore base.DataStore) error {
+	_, err := metadataStore.Exists("fakedoc") // no need to pass a doc and check if it exists, this op checks if the datastore is real
```

Comment:
What's the retry strategy we use for the Exists call?  We don't want to report the metadataStore as missing due to a transient server connectivity issue.  

Would requiring a successful GetCollectionID() be a more reliable alternative?

Reply:
There's no retry on `GetCollectionID` either which sort of runs into the same problem where we do not retry and expect the connection to be open.

`NamedDataStore` calls `Collection.setCollectionID` in `newCollection` constructor. For probably inconsistent reasons, we construct `Collection{}` literally because `collectionID` on default collection is zero.

I think this might expose a different problem where we have transient errors in `NewDatabaseContext` from `bucket.NamedDataStore` since we don't retry KV ops there.

It might be well to put all the retry logic into the `newCollection` constructor which would handle both these circumstances. What retry time is reasonable?

Reply:
In further investigation, this is done for non metadatastore in https://github.com/couchbase/sync_gateway/blob/master/rest/server_context.go#L507

It seems like I should move this logic to `server_context.go` so it works the same for each collection. It returns an `ErrAuthNotFound`.

I would be tempted to have `DefaultDataStore` call `NamedDataStore` but the latter returns and error and the former does not and would need to change sg-bucket/walrus.